### ISSUE TITLE
ARIA improvements for event elements, tabindex

### ISSUE THREAD

**Issue by tyrylu** [fullcalendar/fullcalendar#3364]
*New solution proposed by @arshaw on 8/30/2021:*

Event elements are not always tabbable nor do they always trigger something when the SPACE/ENTER keys are pressed when focused. The solution entails:

1. **Detecting whether the event is "interactable"**. It is interactable if one of the following is true:
  - it has an `event.url` property
  - an `eventClick` handler is defined calendar-wide or for the current view
  - a new `event.interactable` property is set to `true`. This will be useful for users that attach interactions with [eventDidMount](https://fullcalendar.io/docs/event-render-hooks) hook. it must NOT be set to `false`.
2. Is the event interactable?
  - If it IS interactable, we need to ensure it is in the tab order:
    - if it has a `event.url`, no further work is required. An `<a href="">` will be produced, which ensures it is in the tab order
    - otherwise, an `<a>` tag without an href will be produced, which will NOT be in the tab order. Unfortunately, we must keep element as an `<a>`, otherwise it would be a breaking change. Let's put a `tabindex="0"` in tandem with detecting click, ENTER press, and SPACE press.
  - If it IS NOT interactable, we need to put `role="presentation"` on the `<a>` tag to avoid confusing screen readers. In future version of FullCalendar we will move this markup to a different tag (`<span>` ?)

------

*Original request by @tyrylu:*

Hello.
Would it be possible to announce to accessibility technologies that the days and event records may be clicked? It is not obvious, at least in  the usage scenario of Owncloud/Nextcloud. I am not sure what could be done than adding a button role to the elements, you may have some better ideas.

**Comment by arshaw** [comment]
what's the accessibility usecase for making events more buttony? so the user can tab to the event, and press enter to trigger a "click" (that's already done). For [better screen reader support](https://github.com/fullcalendar/fullcalendar/issues/2536) somehow?

**Comment by tyrylu** [comment]
Yes, it is possible to get an interaction using the keyboard, but a screen reader does not indicate that an interaction is possible and it is not sometimes obvious. How would you handle that? I did not have any other idea than the role solution, perhaps only if an event listener is attached (if this is easily doable).

**Comment by arshaw** [comment]
Events that have a `url` will be tabable.. they are `<a>` tags with `href`s. In fact, all rendered events are `<a>` tags.

I'm not sure what technique to use to make each event, regardless of url, tabable. will have to do some research.

**Comment by tyrylu** [comment]
Hi.
When i successfully drilled to the event a elements, i noticed, that they do not have a href attribute. Most screen readers in this case do not announce the element as something interactive and any javascript event handlers don't seem to change anything in this regard. So this is probably the easier part of the fix.

**Comment by advorto-leon-slater** [comment]
As this is still an open issue, it's worth mentioning that, while this isn't currently catered for by fullCalendar, you can use the available callbacks to correct most of it yourself - it's not an elegant solution, but it works for the time being if this is a requirement for you.
e.g.
`
$('#your-element').fullCalendar({
    // other options
    viewRender: function (view, $el) {
        $el.find('a[data-goto]').filter(function () {
            return !$(this).attr('href');
        }).attr('href', '#');
    },
    eventRender: function (event, $el) {
        ($el[0].nodeName === 'A' ? $el : $el.find('a')).attr('href', '#');
    }
});
`

**Comment by mynamesleon** [comment]
As this is still an open issue, it's worth mentioning that, while this isn't currently catered for by fullCalendar, you can use the available callbacks to correct most of it yourself - it's not an elegant solution, but it works for the time being if this is a requirement for you.
e.g.
```javascript
$('#your-element').fullCalendar({
    // other options
    viewRender: function (view, $el) {
        $el.find('a[data-goto]').filter(function () {
            return !$(this).attr('href');
        }).attr('href', '#');
    },
    eventRender: function (event, $el) {
        // if you're doing something when the event is clicked...
        ($el[0].nodeName === 'A' ? $el : $el.find('a')).attr('href', '#');
    }
});
```

**Comment by arshaw** [comment]
this could probably be fixed by adding a `tabindex` to each event element. though it should probably be an option that's turned off by default

**Comment by anupshah** [comment]
Hi, just coming across Full Calendar, and it has some great functionality!

I was looking into the accessibility side of things and found I couldn't keyboard access to most of the events (I was primarily looking at the Time Grid demo on your main web site.)

Regarding this note on the tabindex:

> though it should probably be an option that's turned off by default

I'd like to suggest that maybe it should be on by default? It helps not just those with additional needs (disability), but even power users, to use the keyboard arrow keys etc.

The use case I had in my mind is that you

- Tab into the component (you can already navigate the days, back and forth, today etc via the navigation buttons.
- If you press enter on them, then the first cell gets keyboard focus, and you can use cursor keys to navigate to any cell.
- If a cell has an event in it, then that get focus so you can press enter to open it, for example.
- If you press enter on an empty cell then you can create an event there, instead

So, in effect, enter on the keyboard acts the same as a click event - I think it is just they cannot be tab-focused into? And adding cursor key support just enhances that even more, either for power users or for those using assistive technology

If you use aria `role=grid`, plus various keyboard events to support cursor key navigation, that may give you usability + accessibility enhancements that may also help with screen readers (which I note there is another ticket open for that you linked to earlier)

This has more info: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Grid_Role

(I'll comment on the other screen reader ticket with this same link in case that is helpful.)

**Comment by christianortega123** [comment]
Was anyone able to create a functioning implementation of tabbing through events? Would appreciate any response!

**Comment by gbonker** [comment]
@christianortega123 the following `eventDidMount` block made events tabbable and clickable for my team:

```
var calendar = new FullCalendar.Calendar(calendarEl, {
        eventDidMount: function(info) {
          info.el.href="javascript:void(0);";
        },
});
```

**Comment by arshaw** [comment]
I've put more effort into designing this feature and I've updated the initial post. I will implement this soon, by the end of September.

@anupshah , thanks for your thoughts about arrow keys. I have separate tickets about event drag-n-drop and date selecting:
https://github.com/fullcalendar/fullcalendar/issues/2535
https://github.com/fullcalendar/fullcalendar/issues/6528

**Comment by arshaw** [comment]
The mentioned improvements were implemented and released in [v5.10.0](https://github.com/fullcalendar/fullcalendar/releases/tag/v5.10.0)

The proposed `event.interactable` setting has been named `event.interactive` instead. Also, there is a global flag for overriding this: `eventInteractive`.
